psi4/driver/p4util/exceptions.py

```diff
@@ -81,10 +81,11 @@ class ConvergenceError(PsiException):
     error message *msg* to standard output stream and output file.
 
     """
-    def __init__(self, eqn_description, maxit):
+    def __init__(self, eqn_description, maxit, wfn=None):
```

Comment:
Hmm, is convergence error used elsewhere? Having this as an optional arg could be dangerous when folks are always expecting for this exception to contain a Wfn.

Reply:
I thought it was only used in py-side SCF iterations, so that's a problem for another branch.

Turns out it is used by `diatomic`; that one's off the hook b/c there's no `core.driver` calls involved, so the user has no reasonable expectation of wfn-returnedness.

Also used in `frac`. There, there's simply too many wfns around to single out one to return.

So I think it's ok, subject to fixing in scfitertopy. Other thoughts?

Reply:
Hmm, could we separate out iterative and geometry convergence errors? Could help folks catch the right errors, but subclassing Exceptions can go too far.

Reply:
Sure, though I don't think iterative vs geometry convergence errors will be the separator in the end
```
ConvergenceError (no wfn)
-- OptimizationConvergenceError (wfn) (though keep in mind this could be a dinky wfn if it's from cbs() for example)
-- SCFConvergenceError(wfn) (presently in another branch)
-- -- OscillatingSCFConvergenceError(wfn) (some of these may be internal-only for the smart SCF driver)
-- CCConvergenceError(wfn) (should cc* shift its language boundary)
-- FracConvergenceError(no wfn) (still don't know what to return)
etc.
```

I'm all for a more complete error set now that we can do interesting things with them. Presently still in minimal-working-state exception set.

Would you rather I subclassed out `GeometryConvergenceError` now, so `ConvergenceError` could stay wfn-less? That's fine by me.

Reply:
What about something simpler:
```
ConvergenceError (no wfn)
-- OptimizationConvergenceError (wfn)
-- MethodConvergenceError(wfn) (name?)
```

I dont think we want to go too deep into the individual convergence issues until there is a real need?